node/core/pvf/src/host.rs

```diff
@@ -0,0 +1,1049 @@
+			run_prepare_pool,
+			run_execute_queue,
+			run_sweeper
+		);
+
+		run(
+			Inner {
+				cache_path: config.cache_path,
+				cleanup_pulse_interval: Duration::from_secs(3600),
+				artifact_ttl: Duration::from_secs(3600 * 24),
+				artifacts,
+				to_host_rx,
+				to_prepare_queue_tx,
+				from_prepare_queue_rx,
+				to_execute_queue_tx,
+				to_sweeper_tx,
+				awaiting_prepare: AwaitingPrepare::default(),
+			},
+			run_prepare_pool,
+			run_prepare_queue,
+			run_execute_queue,
+			run_sweeper,
+		)
+		.await
+	};
+
+	(validation_host, run)
+}
+
+/// An execution request that should execute the PVF (known in the context) and send the results
+/// to the given result sender.
+#[derive(Debug)]
+struct PendingExecutionRequest {
+	params: Vec<u8>,
+	result_tx: ResultSender,
+}
+
+/// A mapping from an artifact ID which is in preparation state to the list of pending exeuction
+/// requests that should be executed once the artifact's prepration is finished.
+#[derive(Default)]
+struct AwaitingPrepare(HashMap<ArtifactId, Vec<PendingExecutionRequest>>);
+
+impl AwaitingPrepare {
+	fn add(
+		&mut self,
+		artifact_id: ArtifactId,
+		params: Vec<u8>,
+		result_tx: ResultSender,
+	) {
+		self.0
+			.entry(artifact_id)
+			.or_default()
+			.push(PendingExecutionRequest { params, result_tx });
+	}
+
+	fn take(&mut self, artifact_id: &ArtifactId) -> Vec<PendingExecutionRequest> {
+		self.0.remove(artifact_id).unwrap_or_default()
+	}
+}
+
+struct Inner {
+	cache_path: PathBuf,
+	cleanup_pulse_interval: Duration,
+	artifact_ttl: Duration,
+	artifacts: Artifacts,
+
+	to_host_rx: mpsc::Receiver<ToHost>,
+
+	to_prepare_queue_tx: mpsc::Sender<prepare::ToQueue>,
+	from_prepare_queue_rx: mpsc::UnboundedReceiver<prepare::FromQueue>,
+
+	to_execute_queue_tx: mpsc::Sender<execute::ToQueue>,
+	to_sweeper_tx: mpsc::Sender<PathBuf>,
+
+	awaiting_prepare: AwaitingPrepare,
+}
+
+#[derive(Debug)]
+struct Fatal;
+
+async fn run(
+	Inner {
+		cache_path,
+		cleanup_pulse_interval,
+		artifact_ttl,
+		mut artifacts,
+		to_host_rx,
+		from_prepare_queue_rx,
+		mut to_prepare_queue_tx,
+		mut to_execute_queue_tx,
+		mut to_sweeper_tx,
+		mut awaiting_prepare,
+	}: Inner,
+	prepare_pool: impl Future<Output = ()> + Unpin,
+	prepare_queue: impl Future<Output = ()> + Unpin,
+	execute_queue: impl Future<Output = ()> + Unpin,
+	sweeper: impl Future<Output = ()> + Unpin,
+) {
+	macro_rules! break_if_fatal {
+		($expr:expr) => {
+			match $expr {
+				Err(Fatal) => break,
+				Ok(v) => v,
+			}
+		};
+	}
+
+	let cleanup_pulse = pulse_every(cleanup_pulse_interval).fuse();
+	futures::pin_mut!(cleanup_pulse);
+
+	let mut to_host_rx = to_host_rx.fuse();
+	let mut from_prepare_queue_rx = from_prepare_queue_rx.fuse();
+
+	// Make sure that the task-futures are fused.
+	let mut prepare_queue = prepare_queue.fuse();
+	let mut prepare_pool = prepare_pool.fuse();
+	let mut execute_queue = execute_queue.fuse();
+	let mut sweeper = sweeper.fuse();
+
+	loop {
+		futures::select_biased! {
```

Comment:
Why biased?

Reply:
Good point, forgot to left a comment. The same reason as in the prep queue: just to enable deterministic testing.